java/ql/src/Likely Bugs/Collections/ReadOnlyContainer.ql

```diff
@@ -40,6 +40,6 @@ where
   ) and
   // Also, any value that `v` is initialized to is a fresh container,
   forall(Expr e | e = v.getAnAssignedValue() | e instanceof FreshContainer) and
-  // and `v` is not implicitly initialized by a for-each loop.
-  not exists(EnhancedForStmt efs | efs.getVariable().getVariable() = v)
+  // and `v` is not implicitly initialized.
+  not v.(LocalVariableDecl).getDeclExpr().hasImplicitInit()
```

Comment:
`v.getAnAssignedValue()` should also account for simple patterns. This makes me more and more likely to  want to simply include the simple patterns in `LocalVariableDeclExpr.getInit` and hence exclude them from `hasImplicitInit`.

Reply:
These are now split so that simple patterns are direct and record patterns are implicit. Note that many of these were already wrong in that they didn't account for catch variables (though that is of course less likely to arise in practice)